### ISSUE TITLE
Complete API types used

### ISSUE THREAD

**Issue by mathstuf** [mgattozzi/github-rs#16]
Things to do:

  - [ ] support all fields
  - [ ] document all fields

Problems that will arise:

`Repo` has a `parent` key which is itself another `Repo`, but this causes a recursive structure. I don't know if serde supports using `Option<Box<Repo>>` instead.

**Comment by mathstuf** [comment]
Hmm, the various `_url` fields can probably be ignored and instead just be done with other methods on the various traits.

**Comment by mgattozzi** [comment]
Generally this is why I made different types even if it was the exact same field. However in this case it might not be possible. I'll have to dig in to Serde to see if that's possible. I'm guessing this is common enough that they have a solution for it.

**Comment by mgattozzi** [comment]
It's been a while but I've completely rewritten the API for an upcoming release. It'll be using Serde's [Value type](https://docs.serde.rs/serde_json/value/enum.Value.html) making it up to the user how they want to go from there. It allows easier access of the JSON and the only thing one would need to refer to is the GitHub docs. @mathstuf would this be sufficient for you in order to close the issue or would you prefer documentation of the in transit types (since it uses a builder pattern) since they are different.

**Comment by mgattozzi** [comment]
Didn't mean to close that....

**Comment by mathstuf** [comment]
Actual structs are nice, but I can work with `Value` types.

**Comment by mgattozzi** [comment]
I could add structs in as an optional feature flag in the future. This gives more flexibility to end users with what they want.

**Comment by mathstuf** [comment]
You'll probably hit the same problem I had with gitlab, but maybe not as badly: if you don't get the types right (e.g., missing that a field can be `Option<T>` rather than just `T`), you're stuck with a major bump because fixing the type is backwards-incompatible.

We solved that by just creating a mapping from `gitlab 8.17` into `rust-gitlab-0.817.0` and not doing a `1.0` until we're more confident that upstream isn't going to break us again in the future. Luckily it's FOSS and we can help fix it when that happens.

**Comment by mgattozzi** [comment]
Yeah I can totally see that happening and it did in prior version. I thought those would be minor bumps? 1.0 -> 1.1 when an enum or something changes rather than a major bump. Or is that for prior to 1.0 with semantic versioning? If it is then this can be made into a 1.0 blocking issue

**Comment by mathstuf** [comment]
If the enum is meant to be "complete", adding variants breaks match blocks without a "catch-all" arm. Also, changing a field into Option is breaking if it is realized that some field can be null.

These are minor bumps before 1.x, but major afterwards.
